File: velocity/src/main/java/net/azisaba/azisabaachievements/velocity/plugin/PluginConfig.java
package net.azisaba.azisabaachievements.velocity.plugin;

import net.azisaba.azisabaachievements.common.sql.DatabaseConfig;
import ninja.leaping.configurate.ConfigurationNode;
import ninja.leaping.configurate.yaml.YAMLConfigurationLoader;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Properties;

public class PluginConfig {
    public final String redisHost;
    public final int redisPort;
    public final String redisUsername;
    public final String redisPassword;
    public final DatabaseConfig databaseConfig;

    public PluginConfig(@NotNull Path configPath) throws IOException {
        if (Files.notExists(configPath)) {
            try (InputStream configStream = VelocityPlugin.class.getClassLoader().getResourceAsStream("/config.yml")) {
                if (configStream != null) {
                    Files.copy(configStream, configPath);
                }
            }
        }
        ConfigurationNode node = YAMLConfigurationLoader.builder().setPath(configPath).build().load();
        this.redisHost = node.getNode("redis", "host").getString("localhost");
        this.redisPort = node.getNode("redis", "port").getInt(6379);
        this.redisUsername = node.getNode("redis", "username").getString();
        this.redisPassword = node.getNode("redis", "password").getString();
        this.databaseConfig = loadDatabaseConfig(node.getNode("database"));
    }

    @Contract("_ -> new")
    private @NotNull DatabaseConfig loadDatabaseConfig(@NotNull ConfigurationNode node) {
        String driver = node.getNode("driver").getString();
        String scheme = node.getNode("scheme").getString("jdbc:mariadb");
        String hostname = node.getNode("hostname").getString("localhost");
        int port = node.getNode("port").getInt(3306);
        String name = node.getNode("name").getString("azisabaachievements");
        String username = node.getNode("username").getString();
        String password = node.getNode("password").getString();
        Properties properties = new Properties();
        node.getNode("properties").getChildrenMap()
                .forEach((key, value) -> properties.setProperty(String.valueOf(key), value.getString()));
        return new DatabaseConfig(driver, scheme, hostname, port, name, username, password, properties);
    }
}

File: cli/src/main/kotlin/net/azisaba/azisabaachievements/cli/config/PluginConfig.kt
package net.azisaba.azisabaachievements.cli.config

import net.azisaba.azisabaachievements.cli.Log
import net.azisaba.azisabaachievements.common.sql.DatabaseConfig
import org.jetbrains.annotations.Contract
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.yaml.YamlConfigurationLoader
import java.nio.file.Files
import java.nio.file.Path
import java.util.Properties
import kotlin.io.path.absolutePathString

class PluginConfig(configPath: Path) {
    val redisHost: String
    val redisPort: Int
    val redisUsername: String?
    val redisPassword: String?
    val databaseConfig: DatabaseConfig

    init {
        Log.debug("Attempting to load config from ${configPath.absolutePathString()}")
        if (Files.notExists(configPath)) {
            PluginConfig::class.java.classLoader.getResourceAsStream("/config.yml").use { configStream ->
                if (configStream != null) {
                    Files.copy(configStream, configPath)
                    Log.info("Created config file at ${configPath.absolutePathString()}")
                } else {
                    Log.warn("Default config.yml is missing")
                }
            }
        }
        val node = YamlConfigurationLoader.builder().path(configPath).build().load()
        redisHost = node.node("redis", "host").getString("localhost")
        redisPort = node.node("redis", "port").getInt(6379)
        redisUsername = node.node("redis", "username").string
        redisPassword = node.node("redis", "password").string
        databaseConfig = loadDatabaseConfig(node.node("database"))
    }

    @Contract("_ -> new")
    private fun loadDatabaseConfig(node: ConfigurationNode): DatabaseConfig {
        val driver = node.node("driver").string
        val scheme = node.node("scheme").getString("jdbc:mariadb")
        val hostname = node.node("hostname").getString("localhost")
        val port = node.node("port").getInt(3306)
        val name = node.node("name").getString("azisabaachievements")
        val username = node.node("username").string
        val password = node.node("password").string
        val properties = Properties()
        node.node("properties")
            .childrenMap()
            .forEach { (key: Any, value: ConfigurationNode) -> properties.setProperty(key.toString(), value.string) }
        return DatabaseConfig(driver, scheme, hostname, port, name, username, password, properties)
    }
}
